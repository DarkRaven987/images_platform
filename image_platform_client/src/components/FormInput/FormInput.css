.form-input > label {
  color: #c9c9c9 !important ;
}

.form-input {
  display: flex;
  flex-direction: column;
}

.form-input > input {
  width: calc(100% - 20px);
  font-size: 18px;
  padding: 8px;
  margin-top: 8px;
}

.form-input .error-message {
  color: red;
  margin: 10px 0 0;
  font-size: 14px;
  max-width: calc(100% - 20px);
}
