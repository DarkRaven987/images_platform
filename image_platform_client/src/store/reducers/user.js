import { createSlice } from '@reduxjs/toolkit';
import { agent } from '../../utils/agent';
import { API_SIGN_IN_URL, API_SIGN_UP_URL } from '../../utils/agentConsts';

const initialState = null;

const { actions, reducer } = createSlice({
  name: 'user',
  initialState,
  reducers: {
    saveUserData(state, action) {
      return (state = action.payload);
    },
    clearUserData(state, action) {
      return (state = action.payload);
    },
  },
});

export const { saveUserData, clearUserData } = actions;

export const signInAction = async ({ dispatch, username, password }) => {
  if (!username || !password) return false;

  try {
    const userData = await agent.post(API_SIGN_IN_URL, { username, password });
    const { user, accessToken, refreshToken } = userData?.data;

    if (!user?.PK) return false;

    dispatch(saveUserData(user));
    localStorage.setItem('user', JSON.stringify(user));
    localStorage.setItem('accessToken', `Bearer ${accessToken}`);
    localStorage.setItem('refreshToken', `Bearer ${refreshToken}`);

    return true;
  } catch (err) {
    console.error(err);
    return false;
  }
};

export const signUpAction = async ({ username, password }) => {
  if (!username || !password) return false;

  try {
    await agent.post(API_SIGN_UP_URL, { username, password });

    return true;
  } catch (err) {
    console.error(err);
    return false;
  }
};

export const signOutAction = () => {
  localStorage.removeItem('user');
  localStorage.removeItem('accessToken');
  localStorage.removeItem('refreshToken');
  window.location.replace(`${window.location.origin}/login`);
};

export default reducer;
