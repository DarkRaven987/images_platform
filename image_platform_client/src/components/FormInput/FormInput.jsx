import { useState, useCallback } from 'react';
import './FormInput.css';

const FormInput = ({
  value = '',
  onChange = () => {},
  type = 'text',
  id,
  label,
  className = '',
  style,
  error,
  refer = {},
}) => {
  const { onChange: referOnChange, ...rest } = refer;

  const [val, setVal] = useState(value);

  const handleChange = useCallback(
    (e) => {
      setVal(e?.target?.value);
      onChange(e.target.value);
      refer?.onChange && refer?.onChange(e);
    },
    [onChange, refer],
  );

  return (
    <div
      className={`form-input ${className}`}
      style={style}
      data-testid="form-input-container"
    >
      <label htmlFor={id} data-testid="form-input-label">
        {label}
      </label>
      <input
        id={id}
        type={type}
        value={val}
        onChange={handleChange}
        data-testid="form-input"
        {...rest}
      />
      {!!error && (
        <p data-testid="form-input-error" className="error-message">
          {error.message}
        </p>
      )}
    </div>
  );
};

export default FormInput;
