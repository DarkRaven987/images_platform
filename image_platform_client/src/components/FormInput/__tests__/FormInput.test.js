import { render, screen, cleanup } from '@testing-library/react';
import user from '@testing-library/user-event';
import FormInput from '../FormInput';

const testingText = 'test text';

describe('FormInput component:', () => {
  beforeEach(() => cleanup());

  it('* renders properly', () => {
    render(<FormInput value={testingText} />);
    const InputContainer = screen.queryByTestId('form-input-container');
    const InputLabel = screen.queryByTestId('form-input-label');
    const Input = screen.queryByTestId('form-input');
    const InputErrors = screen.queryByTestId('form-input-error');

    expect(InputContainer).toBeTruthy();
    expect(InputLabel).toBeTruthy();
    expect(Input).toBeTruthy();
    expect(InputErrors).not.toBeTruthy();

    expect(Input.value).toBe(testingText);
  });

  it('* accepts and applies attributes', () => {
    render(
      <FormInput
        className="custom-class-name"
        id="test-1"
        label="Test Input"
        type="password"
        value=""
      />,
    );
    const InputContainer = screen.queryByTestId('form-input-container');
    const InputLabel = screen.queryByTestId('form-input-label');
    const Input = screen.queryByTestId('form-input');
    const InputErrors = screen.queryByTestId('form-input-error');

    expect(InputContainer.classList).toContain('form-input');
    expect(InputContainer.classList).toContain('custom-class-name');

    expect(InputLabel.textContent).toBe('Test Input');

    expect(Input.type).toBe('password');

    expect(InputErrors).not.toBeTruthy();
  });

  it('* changes the value', async () => {
    user.setup();

    let test = '';

    render(
      <FormInput
        className="custom-class-name"
        id="test-1"
        label="Test Input"
        type="password"
        onChange={(val) => (test = val)}
        value={test}
      />,
    );

    const Input = screen.queryByTestId('form-input');

    await user.type(Input, 'User typed here');

    expect(Input.value).toBe(test);
  });

  it('* displays error if provided', () => {
    render(
      <FormInput
        className="custom-class-name"
        id="test-1"
        label="Test Input"
        type="password"
        error={{
          message: 'Field is required',
        }}
      />,
    );

    const InputErrors = screen.queryByTestId('form-input-error');

    expect(InputErrors).toBeTruthy();

    expect(InputErrors.textContent).toBe('Field is required');
  });

  it('* matches the snapshot', async () => {
    user.setup();

    let test = '';

    render(
      <FormInput
        className="custom-class-name"
        id="test-1"
        label="Test Input"
        onChange={(val) => (test = val)}
        value={test}
      />,
    );
    const InputContainer = screen.queryByTestId('form-input-container');
    const Input = screen.queryByTestId('form-input');

    expect(InputContainer).toMatchSnapshot('Input-before-edit');

    await user.type(Input, 'User typed here');

    expect(InputContainer).toMatchSnapshot('Input-after-edit');
  });
});
