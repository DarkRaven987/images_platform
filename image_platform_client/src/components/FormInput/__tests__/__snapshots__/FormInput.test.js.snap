// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput component: * matches the snapshot: Input-after-edit 1`] = `
<div
  class="form-input custom-class-name"
  data-testid="form-input-container"
>
  <label
    data-testid="form-input-label"
    for="test-1"
  >
    Test Input
  </label>
  <input
    data-testid="form-input"
    id="test-1"
    type="text"
    value="User typed here"
  />
</div>
`;

exports[`FormInput component: * matches the snapshot: Input-before-edit 1`] = `
<div
  class="form-input custom-class-name"
  data-testid="form-input-container"
>
  <label
    data-testid="form-input-label"
    for="test-1"
  >
    Test Input
  </label>
  <input
    data-testid="form-input"
    id="test-1"
    type="text"
    value=""
  />
</div>
`;
